packages/PHPStanStaticTypeMapper/TypeMapper/StaticTypeMapper.php

```diff
@@ -51,13 +52,19 @@ public function mapToPhpParserNode(Type $type, string $typeKind): ?Node
     {
         // special case, for autocomplete of return type
         if ($type instanceof SimpleStaticType) {
-            return new Name(ObjectReference::STATIC);
+            return $this->phpVersionProvider->isAtLeastPhpVersion(PhpVersionFeature::STATIC_RETURN_TYPE)
+                ? new Name(ObjectReference::STATIC)
+                : new Name(ObjectReference::SELF);
         }
 
         if ($type instanceof SelfStaticType) {
             return new Name(ObjectReference::SELF);
         }
 
+        if ($type instanceof ThisType) {
+            return new Name(ObjectReference::SELF);
+        }
```

Comment:
@staabm this will make add `self` to return even on `non-final` class  => `?self` on use `return $this`, so it can downgrade the functionality.

I guess we need another "special fixer rule" to change "self" to "static" when possible (depends on php version)

so this change need more consideration :)